File: SIMINOV-HYBRID-PHONEGAP-LIBRARY-TEMPLATE/src/siminov/hybrid/phonegap/library/template/CredentialManager.java
package siminov.hybrid.phonegap.library.template;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Iterator;

import siminov.connect.resource.Resources;
import siminov.hybrid.phonegap.library.template.model.Credential;
import siminov.orm.exception.SiminovCriticalException;
import siminov.orm.exception.SiminovException;
import siminov.orm.log.Log;
import siminov.orm.utils.ClassUtils;
import siminov.orm.utils.EmptyIterator;

public class CredentialManager {

	private static CredentialManager credentialManager = null;
	
	private Credential activeCredential = null;
	private Resources resources = Resources.getInstance();
	
	private CredentialManager() {
		
	}
	
	public static CredentialManager getInstance() {
		
		if(credentialManager == null) {
			credentialManager = new CredentialManager();
		}
		
		return credentialManager;
	}
	
	
	public boolean isAnyActiveCredential() {

		Credential credential = (Credential) ClassUtils.createClassInstance(Credential.class.getName());
		if(credential == null) {
			Log.error(CredentialManager.class.getName(), "isAnyActiveCredential", "NO USER DEFINED CREDENTIAL.");
			throw new SiminovCriticalException(CredentialManager.class.getName(), "isAnyActiveCredential", "NO USER DEFINED CREDENTIAL.");
		}
		
		
		int activeAccountCount = 0;

		try {
			activeAccountCount = (Integer) credential.count().where(Credential.ACTIVE).equalTo(true).execute();
		} catch(SiminovException se) {
			Log.error(CredentialManager.class.getName(), "isAnyActiveCredential", "SiminovException caught while getting active account count, " + se.getMessage());
			throw new SiminovCriticalException(CredentialManager.class.getName(), "isAnyActiveCredential", "SiminovException caught while getting active account count, " + se.getMessage());
		}


		if(activeAccountCount <= 0) {
			return false;
		} else {
			return true;
		}
	}

	public Credential getActiveCredential() {

		Credential credential = (Credential) ClassUtils.createClassInstance(Credential.class.getName());
		if(credential == null) {
			Log.error(CredentialManager.class.getName(), "isAnyActiveCredential", "NO USER DEFINED CREDENTIAL.");
			throw new SiminovCriticalException(CredentialManager.class.getName(), "isAnyActiveCredential", "NO USER DEFINED CREDENTIAL.");
		}

		
		if(activeCredential != null) {
			return activeCredential;
		}
		
		
		Credential[] credentials = null;
		try {
			credentials = (Credential[]) credential.select().
					where(Credential.ACTIVE).equalTo(true).
					execute();
		} catch(SiminovException se) {
			Log.error(CredentialManager.class.getName(), "getActiveCredential", "SiminovException caught while getting active account, " + se.getMessage());
			throw new SiminovCriticalException(CredentialManager.class.getName(), "getActiveCredential", "SiminovException caught while getting active account, " + se.getMessage());
		}

		if(credentials == null || credentials.length <= 0) {
			Log.error(CredentialManager.class.getName(), "getActiveCredential", "No Account Found.");
			return null;
		}


		if(credentials.length > 1) {
			Log.error(CredentialManager.class.getName(), "getActiveCredential", "More Then One Active Account Found.");
			throw new SiminovCriticalException(CredentialManager.class.getName(), "getActiveCredential", "More Then One Active Account Found.");
		}


		activeCredential = credentials[0];
		return activeCredential;
	}

	public Iterator<Credential> getCredentials() {

		Credential credential = (Credential) ClassUtils.createClassInstance(Credential.class.getName());
		if(credential == null) {
			Log.error(CredentialManager.class.getName(), "isAnyActiveCredential", "NO USER DEFINED CREDENTIAL.");
			throw new SiminovCriticalException(CredentialManager.class.getName(), "isAnyActiveCredential", "NO USER DEFINED CREDENTIAL.");
		}

		
		Credential[] credentials = null;
		try {
			credentials = (Credential[]) credential.select().
					where(Credential.ACTIVE).equalTo(true).
					execute();
		} catch(SiminovException se) {
			Log.error(CredentialManager.class.getName(), "getCredentials", "SiminovException caught while getting active account, " + se.getMessage());
			throw new SiminovCriticalException(CredentialManager.class.getName(), "getCredentials", "SiminovException caught while getting active account, " + se.getMessage());
		}

		if(credentials == null || credentials.length <= 0) {
			Log.debug(CredentialManager.class.getName(), "getCredentials", "No Account Found.");
			return new EmptyIterator<Credential>();
		}


		Collection<Credential> accounts = new ArrayList<Credential>();
		for(int i = 0;i < credentials.length;i++) {
			accounts.add(credentials[i]);
		}

		return accounts.iterator();
	}

	public void setActiveCredential(Credential credential) {
		
	}
}

File: connect-sample/iOS/Siminov/Connect/Sample/CredentialManager.m
//
//  CredentialManager.m
//  connect-sample
//
//  Created by Geetika on 31/05/15.
//  Copyright (c) 2015 Siminov. All rights reserved.
//

#import "CredentialManager.h"
#import "SICClassUtils.h"
#import "SICICountClause.h"
#import "SICISelectClause.h"

static CredentialManager *credentialManager = nil;
Credential *activeCredential = nil;

@implementation CredentialManager

+ (CredentialManager *)getInstance {
    
    if(!credentialManager) {
        credentialManager = [[super allocWithZone:NULL] init];
    }
    
    return credentialManager;
}

- (BOOL)isAnyActiveCredential {
    Credential *credential = (Credential *)[SICClassUtils createClassInstance:NSStringFromClass([Credential class])];
    
    if (credential == nil) {
        [SICLog error:NSStringFromClass([self class]) methodName:@"isAnyActiveCredential" message:@"NO USER DEFINED CREDENTIAL."];
        @throw [[SICSiminovCriticalException alloc]initWithClassName:NSStringFromClass([self class]) methodName:@"isAnyActiveCredential" message:@"NO USER DEFINED CREDENTIAL."];
    }
    
    int activeAccountCount = 0;
    
    @try {
        activeAccountCount = [[[[[credential count] where:ACTIVE] equalTo:@"true"] execute] intValue];
    }
    @catch (SICSiminovException *exception) {
        [SICLog error:NSStringFromClass([self class]) methodName:@"isAnyActiveCredential" message:[NSString stringWithFormat:@"SiminovException caught while getting active account count, %@",[exception getMessage]]];
        @throw [[SICSiminovCriticalException alloc]initWithClassName:NSStringFromClass([self class]) methodName:@"isAnyActiveCredential" message:[NSString stringWithFormat:@"SiminovException caught while getting active account count, %@",[exception getMessage]]];
    }
    
    if (activeAccountCount <= 0) {
        return false;
    } else {
        return true;
    }
}

- (Credential *)getActiveCredential {
    Credential *credential = [SICClassUtils createClassInstance:NSStringFromClass([self class])];
    
    if (credential == nil) {
        [SICLog error:NSStringFromClass([self class]) methodName:@"getActiveCredential" message:@"NO USER DEFINED CREDENTIAL."];
        @throw [[SICSiminovCriticalException alloc]initWithClassName:NSStringFromClass([self class]) methodName:@"getActiveCredential" message:@"NO USER DEFINED CREDENTIAL."];
    }
    
    if (activeCredential != nil) {
        return activeCredential;
    }
    
    NSArray *credentials = nil;
    @try {
        credentials = [[[[credential count] where:ACTIVE] equalTo:@"true"] execute];
    }
    @catch (SICSiminovException *exception) {
        [SICLog error:NSStringFromClass([self class]) methodName:@"isAnyActiveCredential" message:[NSString stringWithFormat:@"SiminovException caught while getting active account count, %@",[exception getMessage]]];
        @throw [[SICSiminovCriticalException alloc]initWithClassName:NSStringFromClass([self class]) methodName:@"isAnyActiveCredential" message:[NSString stringWithFormat:@"SiminovException caught while getting active account count, %@",[exception getMessage]]];
    }
    
    if (credentials == nil || credentials.count <= 0) {
       [SICLog error:NSStringFromClass([self class]) methodName:@"getActiveCredential" message:@"No Account Found."];
    }
    
    if (credentials.count>1) {
        [SICLog error:NSStringFromClass([self class]) methodName:@"getActiveCredential" message:@"More Then One Active Account Found."];
        @throw [[SICSiminovCriticalException alloc]initWithClassName:NSStringFromClass([self class]) methodName:@"getActiveCredential" message:@"More Then One Active Account Found."];
    }
    
    activeCredential = credentials[0];
    return activeCredential;
}

- (NSEnumerator *)getCredentails {
    Credential *credential = [SICClassUtils createClassInstance:NSStringFromClass([Credential class])];
   
    if (credential == nil) {
        [SICLog error:NSStringFromClass([self class]) methodName:@"getActiveCredential" message:@"NO USER DEFINED CREDENTIAL."];
        @throw [[SICSiminovCriticalException alloc]initWithClassName:NSStringFromClass([self class]) methodName:@"getActiveCredential" message:@"NO USER DEFINED CREDENTIAL."];
    }
    
    NSArray *credentials = nil;
    @try {
        credentials = [[[[credential select] where:ACTIVE] equalTo:@"true"] execute];
    }
    @catch (SICSiminovException *exception) {
        [SICLog error:NSStringFromClass([self class]) methodName:@"getCredentials" message:[NSString stringWithFormat:@"SiminovException caught while getting active account count, %@",[exception getMessage]]];
        @throw [[SICSiminovCriticalException alloc]initWithClassName:NSStringFromClass([self class]) methodName:@"getCredentials" message:[NSString stringWithFormat:@"SiminovException caught while getting active account count, %@",[exception getMessage]]];
    }
    
    if (credentials == nil || credentials.count <= 0) {
        [SICLog error:NSStringFromClass([self class]) methodName:@"getCredentials" message:@"No Account Found."];
        return [[NSEnumerator alloc]init];
    }
    
    NSMutableArray *accounts = [[NSMutableArray alloc]init];
    for (int i = 0; i<[credentials count]; i++) {
        [accounts addObject:credentials[i]];
    }
    
    return [accounts objectEnumerator];
}

- (void)setActiveCredential:(Credential *)credential {

}


@end
